### PR TITLE
[OP] broadcast plus, minus, mul, ...

### PR DESCRIPTION
**Pull request by mli**
add a new op `_broadcast_plus`, both forward and backward are tested. 

before extending to other elemental binary operators easily, i'd like to PR first because the implementation is unexpected complicate, i may be get something wrong..

### CODE REVIEW

File: src/operator/elementwise_binary_broadcast_op-inl.h [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+          }
+        } else {
+          LOG(FATAL) << "no broadcast is needed";
+        }
+      }
+    });
+}
+
+template<typename xpu>
+void PlusBroadcastBackward_(const OutputGrad& out_grad,
```
Comment (piiswrong) [inline src/operator/elementwise_binary_broadcast_op-inl.h]: can we reuse backward code for different operation?

File: python/mxnet/symbol.py [hunk 2]
```diff
@@ -35,8 +35,52 @@ def __init__(self, handle):
...
+        """add with broadcast
+
+        Examples
+        --------
+        >>> a = mx.sym.Variable('a')
+        >>> b = mx.sym.Variable('b')
+        >>> c = a + b
+
+        >>> dev = mx.cpu();
+        >>> x = c.bind(dev, args={'a': mx.nd.ones((2,2)), 'b' : mx.nd.ones((2,2))})
```
Comment (piiswrong) [inline python/mxnet/symbol.py]: can you do it as simple binary op, so that you can also use this for imperative?

**Comment by pluskid** [comment]
Very nice! I would suggest an add on to our shape inference mechanism here: we should allow the user to attach shape information to a `Variable`. During shape inference, this information can be used. This is especially useful in the case when broadcasting is supported. Sometimes I have a parameter that should be broadcast through the samples in a mini-batch, if I use broad cast plus to construct a symbol, the shape inference will not be able to figure out the shape of that parameter purely based on the data shape.

**Comment by mli** [comment]
@pluskid i didn't get your question. currently the shape inference is done by `BinaryBroadcastShape_`, which will determine the broadcast dimensions. One can also always use `symbol.Reshape` to make the broadcast performs correctly.

File: src/operator/elementwise_binary_broadcast_op-inl.h (hunk 1 above)
Comment (mli) [inline src/operator/elementwise_binary_broadcast_op-inl.h]: only plus and minus can reuse the same codes

**Comment by pluskid** [comment]
@mli For example, this currently works
```python
import mxnet as mx

a = mx.sym.Variable('a')
b = mx.sym.Variable('b')
c = a+b
c.infer_shape(a=(10,20))
```

But if we replace `+` with boradcasting plus, will it still work? How is it going to infer the shape of `b` purely based on `a` for broadcasting? If you need a real scenario, there is this full LSTM implementation in acoustic model where a broadcasting elementwise multiplication is needed.

**Comment by mli** [comment]
the only way works is 

```python
>>> d.infer_shape(a=(10,20), b=(10,1))
>>> d = mx.sym.BroadcastPlus(a,b)
```
so your use case is that the shape of `b` is unclear?

**Comment by pluskid** [comment]
@mli in my use case, `b` is a parameter, while `a` is the data (or some hidden layer computed from the data). Our model or module are all based on the assumption that shape information are only from `provide_data` and `provide_label` from a `DataIter`, everything else should be able to be shape inferred. This works very well so far, except for this case. I think allowing the user to attach shape information to `Variable` sounds like a simple solution.

```python
a = mx.sym.Variable('a')
b = mx.sym.Variable('b', shape=(1, 10))
c = mx.sym.BroadcastPlus(a,b)

c.simple_bind(a=(5, 10))
```

**Comment by mli** [comment]
understood now. i'm agree that adding a shape attribution is cleaner than the possible solution that passes additional shape info to the data iterator

**Comment by piiswrong** [comment]
I guess dont broadcast unless you have to and chiyuan's example still works.

**Comment by pluskid** [comment]
@piiswrong My problem is that I know that the expected behavior is to be shared. For example, if I want to implement the fully connected layer with low level arithmetics, I could say `out = matmul(X, W) + b`, here `b` should not be of shape `N-by-d`, instead it should be broadcast and have the shape `1-by-d` or just `(d, )`.

File: src/operator/elementwise_binary_broadcast_op-inl.h [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+          }
+        } else {
+          LOG(FATAL) << "no broadcast is needed";
+        }
+      }
+    });
+}
+
+
+template<typename xpu, typename LHS_OP, typename RHS_OP>
```
Comment (piiswrong) [inline src/operator/elementwise_binary_broadcast_op-inl.h]: if you feed both left hand right to both LHD_OP and RHS_OP mulbackward can be merged right? This shouldn't increase time or memory.

**Comment by tqchen** [comment]
please check why the python test fails. Note that if bcast do not support inplace, inplace declaration should be disabled

File: src/operator/elementwise_binary_broadcast_op-inl.h [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+MXNET_REGISTER_SIMPLE_OP(_broadcast_minus, XPU)
+.set_symbol_op_name("BroadcastMinus")
+.set_shape_function(BinaryBroadcastShape_)
+.set_function(XPU::kDevMask, BinaryBroadcastForward_<XPU, mshadow::op::minus>, kInplaceLhsOut)
+.set_gradient(XPU::kDevMask, BinaryBroadcastBackward_<
+              XPU, mshadow_op::identity, mshadow_op::negation>, kInplaceOutLhs)
+.describe("lhs minus rhs with broadcast");
+
+MXNET_REGISTER_SIMPLE_OP(_broadcast_mul, XPU)
+.set_symbol_op_name("BroadcastMul")
```
Comment (tqchen) [inline src/operator/elementwise_binary_broadcast_op-inl.h]: for new operators, let us keep the name consistent, with lower cases, so mx.sym and mx.nd have exact the same function.

**Comment by piiswrong** [comment]
@mli Any updates on this?

**Comment by sxjscience** [comment]
Really looking forward to some updates.

**Comment by piiswrong** [comment]
@mli is probably busy lately. Can someone takeover this?

**Comment by piiswrong** [comment]
closing this since we are moving to new one

**Pull request closed without merging**